### ISSUE TITLE
Integrating to ChainAuthFilter

### ISSUE THREAD

**Issue by vycoder** [ahus1/keycloak-dropwizard-integration#96]
Hi, first of all, thank you very much for this nifty little library!

I'm trying to add keycloak from an existing REST API but I can't seem to figure out how to do it. I've tried the following on the docs: 
```java
KeycloakDeployment keycloakDeployment =
        KeycloakDeploymentBuilder.build(configuration.getKeycloakConfiguration());
JaxrsBearerTokenFilterImpl filter = new DropwizardBearerTokenFilterImpl(keycloakDeployment);
environment.jersey().register(filter);
```
But it's deprecated. I can't use `KeycloakBundle` because I need to do a quick look up on a db to check for the user, can't add a custom authenticator within `initialize`. I ended up just referencing KeycloakBundle.java's code and then registering my filter along with my custom authenticator and authorizer within dropwizard's `run()`.

I'm getting the following error:
```bash
ERROR [2021-04-30 01:53:08,359] io.dropwizard.jersey.errors.LoggingExceptionMapper: Error handling a request: b0e2fcf096548546
! java.lang.NullPointerException: null
! at org.eclipse.jetty.server.Request.getBaseRequest(Request.java:169)
! at de.ahus1.keycloak.dropwizard.KeycloakAuthFilter.validateRequest(KeycloakAuthFilter.java:97)
! at de.ahus1.keycloak.dropwizard.KeycloakAuthFilter.filter(KeycloakAuthFilter.java:54)
! at org.glassfish.jersey.server.ContainerFilteringStage.apply(ContainerFilteringStage.java:132)
! at org.glassfish.jersey.server.ContainerFilteringStage.apply(ContainerFilteringStage.java:68)
! at org.glassfish.jersey.process.internal.Stages.process(Stages.java:197)
! at org.glassfish.jersey.server.ServerRuntime$2.run(ServerRuntime.java:318)
! at org.glassfish.jersey.internal.Errors$1.call(Errors.java:271)
! at org.glassfish.jersey.internal.Errors$1.call(Errors.java:267)
! at org.glassfish.jersey.internal.Errors.process(Errors.java:315)
! at org.glassfish.jersey.internal.Errors.process(Errors.java:297)
! at org.glassfish.jersey.internal.Errors.process(Errors.java:267)
! at org.glassfish.jersey.process.internal.RequestScope.runInScope(RequestScope.java:317)
! at org.glassfish.jersey.server.ServerRuntime.process(ServerRuntime.java:305)
! at org.glassfish.jersey.server.ApplicationHandler.handle(ApplicationHandler.java:1154)
! at org.glassfish.jersey.servlet.WebComponent.serviceImpl(WebComponent.java:473)
! at org.glassfish.jersey.servlet.WebComponent.service(WebComponent.java:427)
! at org.glassfish.jersey.servlet.ServletContainer.service(ServletContainer.java:388)
! at org.glassfish.jersey.servlet.ServletContainer.service(ServletContainer.java:341)
! at org.glassfish.jersey.servlet.ServletContainer.service(ServletContainer.java:228)
! at io.dropwizard.jetty.NonblockingServletHolder.handle(NonblockingServletHolder.java:49)
! at org.eclipse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1623)
! at io.dropwizard.servlets.ThreadNameFilter.doFilter(ThreadNameFilter.java:35)
! at org.eclipse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1610)
! at io.dropwizard.jersey.filter.AllowedMethodsFilter.handle(AllowedMethodsFilter.java:45)
! at io.dropwizard.jersey.filter.AllowedMethodsFilter.doFilter(AllowedMethodsFilter.java:39)
! at org.eclipse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1610)
! at org.eclipse.jetty.servlets.CrossOriginFilter.handle(CrossOriginFilter.java:311)
! at org.eclipse.jetty.servlets.CrossOriginFilter.doFilter(CrossOriginFilter.java:265)
! at org.eclipse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1610)
! at org.eclipse.jetty.servlet.ServletHandler.doHandle(ServletHandler.java:540)
! at org.eclipse.jetty.server.handler.ScopedHandler.handle(ScopedHandler.java:146)
! at org.eclipse.jetty.security.SecurityHandler.handle(SecurityHandler.java:524)
! at org.eclipse.jetty.server.handler.HandlerWrapper.handle(HandlerWrapper.java:132)
! at org.eclipse.jetty.server.handler.ScopedHandler.nextHandle(ScopedHandler.java:257)
! at org.eclipse.jetty.server.handler.ContextHandler.doHandle(ContextHandler.java:1345)
! at org.eclipse.jetty.server.handler.ScopedHandler.nextScope(ScopedHandler.java:203)
! at org.eclipse.jetty.servlet.ServletHandler.doScope(ServletHandler.java:480)
! at org.eclipse.jetty.server.handler.ScopedHandler.nextScope(ScopedHandler.java:201)
! at org.eclipse.jetty.server.handler.ContextHandler.doScope(ContextHandler.java:1247)
! at org.eclipse.jetty.server.handler.ScopedHandler.handle(ScopedHandler.java:144)
! at org.eclipse.jetty.server.handler.HandlerWrapper.handle(HandlerWrapper.java:132)
! at com.codahale.metrics.jetty9.InstrumentedHandler.handle(InstrumentedHandler.java:239)
! at io.dropwizard.jetty.RoutingHandler.handle(RoutingHandler.java:52)
! at org.eclipse.jetty.server.handler.gzip.GzipHandler.handle(GzipHandler.java:753)
! at io.dropwizard.jetty.BiDiGzipHandler.handle(BiDiGzipHandler.java:67)
! at org.eclipse.jetty.server.handler.RequestLogHandler.handle(RequestLogHandler.java:56)
! at org.eclipse.jetty.server.handler.StatisticsHandler.handle(StatisticsHandler.java:174)
! at org.eclipse.jetty.server.handler.HandlerWrapper.handle(HandlerWrapper.java:132)
! at org.eclipse.jetty.server.Server.handle(Server.java:505)
! at org.eclipse.jetty.server.HttpChannel.handle(HttpChannel.java:370)
! at org.eclipse.jetty.server.HttpConnection.onFillable(HttpConnection.java:267)
! at org.eclipse.jetty.io.AbstractConnection$ReadCallback.succeeded(AbstractConnection.java:305)
! at org.eclipse.jetty.io.FillInterest.fillable(FillInterest.java:103)
! at org.eclipse.jetty.io.ssl.SslConnection$DecryptedEndPoint.onFillable(SslConnection.java:427)
! at org.eclipse.jetty.io.ssl.SslConnection.onFillable(SslConnection.java:321)
! at org.eclipse.jetty.io.ssl.SslConnection$2.succeeded(SslConnection.java:159)
! at org.eclipse.jetty.io.FillInterest.fillable(FillInterest.java:103)
! at org.eclipse.jetty.io.ChannelEndPoint$2.run(ChannelEndPoint.java:117)
! at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.runTask(EatWhatYouKill.java:333)
! at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.doProduce(EatWhatYouKill.java:310)
! at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.tryProduce(EatWhatYouKill.java:168)
! at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.run(EatWhatYouKill.java:126)
! at org.eclipse.jetty.util.thread.ReservedThreadExecutor$ReservedThread.run(ReservedThreadExecutor.java:366)
! at org.eclipse.jetty.util.thread.QueuedThreadPool.runJob(QueuedThreadPool.java:698)
! at org.eclipse.jetty.util.thread.QueuedThreadPool$Runner.run(QueuedThreadPool.java:804)
! at java.base/java.lang.Thread.run(Thread.java:829)
```

The system has an existing basic jwt authentication, I added the my keycloak filter along with it in `ChainAuthFilter` but It doesn`t seem to work, I'm getting the same error. I tried removing the existing jwt auth, removing ChainAuthFilter to see if it will still work, but it still doesn't.

Any ideas?

**Comment by vycoder** [comment]
Sorry, I somehow made it work. It turns out I forgot to register the security handler. But now my problem is that, the current jwt auth is not working, if keycloak is added to the security handler.

Is there anyway to add the a keycloak filter without interfering with the existing auth strategy?

**Comment by ahus1** [comment]
JaxrsBearerTokenFilterImpl is deprecated, but only in a way that it is looking for a maintainer. If it will be released, you can pick these few classes and maintain them as part of your project. 

Reading the rest of your post, I'm not sure what you want to look up in the database. Usually a lot of information is part of the token. The suggested way with this library is described in the comment: 

```
/* OPTIONAL: override getUserClass(), createAuthorizer() and createAuthenticator() if you want to use
     * a class other than de.ahus1.keycloak.dropwizard.User to be injected by @Auth */
```

This could provide additional methods that read additional information from the token. It might also be possible to look up data from the database, although I've never done that.

**Comment by vycoder** [comment]
Hi @ahus1 thanks for the quick response! 

A little bit of context, I'm maintaining a fairly old REST api and it originally uses a custom JWT authentication strategy to handle user auth. Based on a bearer token, it uses it's own `Principal` using a custom `Authorizer` it looks up the db, and attaches the user object on it. Every resource will then have access to it via `@Auth AuthUser`. (Not how I would do it, but it's the one that I have to work with).

Now here comes keycloak. We're expecting a pool of users from a different platform that uses keycloak to be able use our own system. I was tasked to integrate keycloak within the REST api in a way that the original JWT auth would still work. Basically, I would expect the a bearer token, that would work for either keycloak auth or the existing custom JWT auth. The database is seeded with the new users from the third-party and I'll use the email on the keycloak token to fetch the user record and attach it to our principal.

(Also note that, we're using `@RolesAllowed` annotation, but for all the _keycloak users_, I default them to a basic role implemented on my custom `Authenticator`.)

I was somewhat able to accomplish this just recently. I didn't end up using `JaxrsBearerTokenFilterImpl` because, according to the docs, if I understand correctly, everything will be secured. And I can't figure out a way to make it work while providing a custom authenticator.

What I ended up doing is:
```java
KeycloakAuthFilter<AuthUser> filter = new KeycloakAuthFilter.Builder<AuthUser>()
          .setConfig(config.getKeycloakConfiguration())
          .setAuthenticator(new CustomAuthenticator(config.getKeycloakConfiguration(), userDao))
          .setAuthorizer(new CustomAuthorizer())
          .setRealm(config.getKeycloakConfiguration().getRealm())
          .buildAuthFilter();
```
I'm reusing the same `Authenticator` and some form of an adapter for the `Authorizer` from the existing JWT auth. That worked perfectly, but this piece of code is what stumped me:

```java
KeycloakJettyAuthenticator keycloak = new KeycloakDropwizardAuthenticator();
keycloak.setAdapterConfig(getKeycloakConfiguration(configuration));
ConstraintSecurityHandler securityHandler = new ConstraintSecurityHandler();
environment.getApplicationContext().setSecurityHandler(securityHandler);
environment.getApplicationContext().getSecurityHandler().setAuthenticator(keycloak);
```
I'm not very familiar with `jetty` but it appears that adding a security handler on the the application context is global. It's being applied to everything. But `KeycloakAuthFilter` depends on the request that's being added by `KeycloakDropwizardAuthenticator`, I assume that it's the one that also validates the token. And of course, the existing token doesn't match the token headers that keycloak expects and the whole thing blows up in my face.

It works fine if the only filter is keycloak but not together with the custom JWT Auth that we have.

The workaround that my limited knowledge was able to come up with is to create an Adapter for `KeycloakDropwizardAuthenticator`. It checks whether the token header has a `kid` property as rudimentary differentiation between our existing token and keycloak's. 

`AuthenticatorAdapter.java`
```java
@Override
  public Authentication validateRequest(ServletRequest request, ServletResponse response, boolean mandatory) throws ServerAuthException {
    HttpServletRequest req = ((HttpServletRequest) request);

    Optional<String> authorization = Optional.ofNullable(req.getHeader("Authorization"));
    if (!authorization.isPresent()) {
      return Authentication.NOT_CHECKED;
    }

    // Extract the token, remove the Prefix
    String token = authorization.get()
      .replaceFirst("Bearer", "")
      .replaceFirst("bearer", "")
      .trim();

    String[] chunks = token.split("\\.");
    if (chunks.length < 1) {
      return Authentication.NOT_CHECKED;
    }

    // Keycloak tokens always have "kid" in the header. That's what we'll use to differentiate
    String tokenHeader = new String(Base64.getDecoder().decode(chunks[0]));
    return tokenHeader.contains("kid")
      ? this.keycloakAuthenticator.validateRequest(request, response, mandatory)
      : Authentication.NOT_CHECKED;
  }
```  
I delegate the validate to `KeycloakDropwizardAuthenticator` and if not just issue a no-op using `Authentication.NOT_CHECKED`. 

It works for now, I don't have to change anything on my resource classes, since both `AuthFilters` uses the same custom User class principal.

I'm wondering if there's a better way around this, it seems that keycloak wasn't intended to be used like this, that's why I can't find any example in the documentation regarding this. 

Is my workaround a right approach? Is there a better way to handle this scenario?

**Comment by ahus1** [comment]
I wouldn't call it a workaround. If it works for you and is maintainable, it is a proper solution. Your `validateRequest()` variant seems to be limited to the cases you want to support (that is: bearer only). I'm always unsure with the `Authentication` return codes. Once your happy path and non-happy path test cases pass this should be fine. 

For security related code I opt for checks to be as restrictive as possible. Looking at the code I'd suggest a case insensitive comparison for `Bearer` instead of checking two variants. The `tokenHeader` is actually JSON, so you could decode the JSON before looking for a `kid` key.

**Comment by vycoder** [comment]
Thanks for the inputs! Will take those into consideration. Yes, the `Authentication` return codes were also my worries, I just assumed that `NOT_CHECKED` would be equivalent to not having a `validateRequest` at all, since the existing JWT filter does not use it and only relies on it's `AuthFilter` implementation.

And if ever someone was in the same shoes as I am, I didn't end up using `ChainAuthFilter` after all, since I'm operating on the same header bearer token and not a different Auth strategy altogether. I ended up implementing a composite adapter AuthFilter of some sort. Wherein I'm checking the token header for the `kid` and executing the appropriate `AuthFilter`. This way, only a single `AuthFilter` will get executed, it's either the existing JWT or keycloak's.

Tested things, and so far everything works as expected.

Anyway, thanks for taking time looking into my concern. Closing this issue now. Thanks again!
